File: 0x0018_24.Swap_Nodes_In_Pairs/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode newHead=head;
        if (head!=null && head.next!=null) {
            newHead=head.next;
        } else {
            return head;
        }
        swap(null,head);
        return newHead;
    }
    
    private void swap(ListNode prev,ListNode node) {
        if (node==null) {
            prev.next=null; 
            return;
        }
        if (node.next==null) {
            prev.next=node;
            return;
        }
        ListNode next=node.next;
        if (prev!=null) {
            prev.next=next;
            //System.out.println(prev.val);
        }
        ListNode temp=next.next;
        next.next=node;
        swap(node,temp);
        return;
    }
}

File: 0x0018_24.Swap_Nodes_In_Pairs/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head

        fakehead = ListNode(0)
        prev = fakehead
        cur = head
        while True:
            if cur and cur.next:
                n1 = cur
                n2 = cur.next
                n3 = cur.next.next
                prev.next = n2
                n2.next = n1
                n1.next = None
                prev = n1
                cur = n3
            else:
                if cur:
                    prev.next = cur
                break
        return fakehead.next
